File: leetCode_Python/prob496.py
#!/usr/bin/env python3
#-*- coding:utf-8 -*-

# author    : lambda RDG
# Date&Time : 2019/04/01 14:41:48

"模块的文档注释"
# Name: 496. Next Greater Element I
# Website: https://leetcode.com/problems/next-greater-element-i/
# Description:
import sys

class Solution:
    def nextGreaterElement(self, nums1: List[int], nums2: List[int]) -> List[int]:
        idxMap = {}
        for i, v in enumerate(nums2):
            idxMap[v] = i
        res = []
        for v in nums1:
            index = idxMap[v]
            find = -1
            for v2 in nums2[index:]:
                if v2 > v :
                    find = v2
                    break
            res.append(find)
        return res
        


if __name__ == "__main__":
    pass

File: leetcode/prob496.cpp

#include <iostream>
#include <algorithm>
#include <string>
#include <vector>
#include <map>
#include <utility>
#include <iterator>

#include <cctype>

using namespace std;
/*
* Name: 496. Next Greater Element I
* WebSite: https://leetcode.com/problems/next-greater-element-i/
* Description: 
*
*/
class Solution {
public:
    /*
    * Runtime:    ms
    * Your runtime beats  % of cpp submissions. 
    */
    vector<int> nextGreaterElement(vector<int>& nums1, vector<int>& nums2) {
        std::unordered_map<int, int> valueGreater = {};
        std::vector<int> stack = {};
        for (int v : nums2) {
            while (!stack.empty() && v > stack.back()) {
                valueGreater[stack.back()] = v;
                stack.pop_back();
            }
            stack.push_back(v);
        }
        std::vector<int> res = {};
        for (int v : nums1) {
            int greater = valueGreater[v];
            if (greater == 0)
                res.emplace_back(-1);
            else 
                res.emplace_back(greater);
        }
        return res;
    }
};

int main() 
{
    Solution s; 


    std::cout << "method 1: \n";

    std::cout << std::endl;
    return 0;
}                